alts/src/main/java/io/grpc/alts/internal/TsiHandshakerFactory.java

```diff
@@ -16,11 +16,12 @@
 
 package io.grpc.alts.internal;
 
+import io.grpc.Channel;
 import javax.annotation.Nullable;
 
 /** Factory that manufactures instances of {@link TsiHandshaker}. */
 public interface TsiHandshakerFactory {
 
   /** Creates a new handshaker. */
-  TsiHandshaker newHandshaker(@Nullable String authority);
+  TsiHandshaker newHandshaker(@Nullable Channel handshakerChannel, @Nullable String authority);
```

Comment:
I fear this may have internal repercussions.

Reply:
It is not an issue. I have done it before. Let me prepare an import patch.

Reply:
This looks wrong. The reason authority is passed is because it could be different every time. This channel is also specific to one handshaker implementation.

Instead, you should pass some object when creating the TsiHandshakerFactory. Problably an ObjectPool.